### PR TITLE
Make find & replace functional for keyboard users.

### LINKED ISSUE #8055: Find & Replace is not fully functional for keyboard users.
### Bug Description:

In both Theia and VS Code pressing enter in the focused input(where the new value is entered in find and replace) does not replace the contents I have to manually click on the small icon next to the input. I was expecting it to work when the enter key is pressed.

Peek 2020-06-19 17-22

To me, it looks like a serious accessibility issue for Keyboard users and generally, it would be great if the whole Theia IDE is accessible for those users who only use a keyboard and can't use a mouse.

### Steps to Reproduce:

1. Open a Gitpod workspace.
2. Try to find and replace some text via the Keyboard.

### Additional Information

- Operating System:
- Theia Version:

### PR DESCRIPTION
**Pull request by nisarhassan12**
Fixes #8055

Signed-off-by: Nisar Hassan Naqvi <syednisarhassan12@gmail.com>

#### What it does

Make find & replace functional for keyboard users.

#### How to test

Try find and replace in Theia from sources and try using it with Keyboard.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/search-in-workspace/src/browser/search-in-workspace-widget.tsx [hunk 1]
```diff
@@ -316,19 +316,30 @@ export class SearchInWorkspaceWidget extends BaseWidget implements StatefulWidge
...
         return <div
             title='Toggle Replace'
             className='replace-toggle'
             tabIndex={0}
-            onClick={e => {
-                const elArr = document.getElementsByClassName('replace-toggle');
-                if (elArr && elArr.length > 0) {
-                    (elArr[0] as HTMLElement).focus();
+            onClick={() => {
+                replaceToggle();
```
Comment (akosyakov) [inline packages/search-in-workspace/src/browser/search-in-workspace-widget.tsx:335]: https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#no-bind-fn-in-event-handlers please review other places

**Comment by kittaakos** [comment]
> #### How to test
> Try find and replace in Theia from sources and try using it with Keyboard.

@nisarhassan12, please be a bit more specific. I could use the new feature but I am not sure whether that was the expected behavior or not.

So here is a remark on the [issue](https://github.com/eclipse-theia/theia/issues/8055#issue-641953169):
> In both Theia and VS Code pressing enter in the focused input(where the new value is entered in find and replace) does not replace the contents I have to manually click on the small icon next to the input. I was expecting it to work when the enter key is pressed.

It is not true, the keyboard navigation just works perfectly in VS Code.

screencast 2020-06-24 16-13-56

I have noticed that the `Tab` navigation slightly differs from VS Code; from the `Replace` input the focus should go to the `Match Case` and not directly to `Replace All`  Can we align Theia?

screencast 2020-06-24 16-28-58

File: packages/search-in-workspace/src/browser/search-in-workspace-widget.tsx [hunk 2]
```diff
@@ -316,19 +316,30 @@ export class SearchInWorkspaceWidget extends BaseWidget implements StatefulWidge
...
+        };
+
         return <div
             title='Toggle Replace'
             className='replace-toggle'
             tabIndex={0}
-            onClick={e => {
-                const elArr = document.getElementsByClassName('replace-toggle');
-                if (elArr && elArr.length > 0) {
-                    (elArr[0] as HTMLElement).focus();
```
Comment (kittaakos) [inline packages/search-in-workspace/src/browser/search-in-workspace-widget.tsx:326]: Let's clean this up:
```ts
export class SearchInWorkspaceWidget extends BaseWidget implements StatefulWidget {
  // Do not touch it, use react to modify the DOM
  protected replaceToggleRef: HTMLElement | undefined;
  protected setReplaceToggleRef = (element: HTMLElement | null) => {
    if (element) this.replaceToggleRef = element;
  };

  // enhance the replace toggle
  protected renderReplaceFieldToggle(): React.ReactNode {
    // other stuff
    return <div
      title='Toggle Replace'
      className='replace-toggle'
      ref={this.setReplaceToggleRef} // <-- this is the new thing
      tabIndex={0}
      ...
  }
}
```

And then, you can get rid of:
```ts
const elArr = document.getElementsByClassName('replace-toggle');
if (elArr && elArr.length > 0) {
  (elArr[0] as HTMLElement).focus();
}
```

And replace it with:
```ts
if (this.replaceToggleRef) {
  replaceToggleRef.focus();
}
```

File: packages/search-in-workspace/src/browser/search-in-workspace-widget.tsx [hunk 3]
```diff
@@ -316,19 +316,30 @@ export class SearchInWorkspaceWidget extends BaseWidget implements StatefulWidge
...
+            onClick={() => {
+                replaceToggle();
+            }}
+            onKeyPress={e => {
+                if (e.key === 'Enter' || e.key === ' ') {
```
Comment (kittaakos) [inline packages/search-in-workspace/src/browser/search-in-workspace-widget.tsx:338]: What is `e.key === ' '`? What issue are you trying to solve here?

**Comment by kittaakos** [comment]
> slightly differs from VS Code

> Can we align Theia?

Let's leave as is. The rule of thumb is to mimic VS Code if possible, but in this case, the VS Code behavior is counterintuitive. In Code, the `controls` HTML element (for the match case, match whole world, and the regexp match commands) is not contained in the HTML element for the _Find_ `input`. They're siblings, so the order of the DOM traversal differs.

File: packages/search-in-workspace/src/browser/search-in-workspace-widget.tsx (hunk 3 above)
Comment (nisarhassan12) [reply packages/search-in-workspace/src/browser/search-in-workspace-widget.tsx]: Thanks! @kittaakos here I'm making sure that the users can use the `Space` key as well to do what they can do with `Enter` as  both `Enter` and `Space` keys can be used interchangeably to activate buttons.
 
Source: https://webaim.org/techniques/keyboard/

image

**Comment by nisarhassan12** [comment]
@kittaakos @akosyakov many thanks for the review I have made the requested changes.

File: packages/search-in-workspace/src/browser/search-in-workspace-widget.tsx [hunk 4]
```diff
@@ -435,15 +452,27 @@ export class SearchInWorkspaceWidget extends BaseWidget implements StatefulWidge
...
                 title='Replace All'
+                tabIndex={0}
                 className={`replace-all-button${enabled ? ' ' : ' disabled'}`}
                 onClick={() => {
-                    if (enabled) {
-                        this.resultTreeWidget.replace(undefined);
+                    replace();
+                }}
+                onKeyPress={e => {
+                    if (e.key === 'Enter') {
```
Comment (nisarhassan12) [inline packages/search-in-workspace/src/browser/search-in-workspace-widget.tsx:471]: @akosyakov I should have also added support for  `Space` key here as mentioned in https://github.com/eclipse-theia/theia/pull/8079#discussion_r446644852 but there is an issue shown in the below Gif(sorry I don't know how to record which key is pressed):

space

Using Space key instantly closes the prompt and does the replace all automatically without any action being done on the confirmation prompt. I will look into fixing this probably in the next week.

**Comment by vince-fugnitto** [comment]
@nisarhassan12 do you still plan on working on the pull-request, the last update was over 2 years ago. If not I think it is fine to close the pull-request.

**Comment by vince-fugnitto** [comment]
Closing due to inactivity, we can always revisit the pull-request later as it is still part of the repo's history.

**Pull request closed without merging**